[WIP] vispy.plot - Colorbar

bollu opened the pull request:
Checklist:
- [x] add a `colorbar` function to `PlotWidget`
- [x] separate out border rendering into `BorderVisual`
- [x] integrate border rendering to `ViewBox`
- [ ] fix coordinate systems for `ColorBar`
- [ ] fix text overlapping
- [ ] convert it into a `PlotWidget`
- [ ] write docs on the `colorbar` method
- [ ] make an example

bollu committed:
get a colorbar into the PlotWidget

bollu commented:
@Eric89GXL: ping, so you receive future updates

larsoner commented:
Hmm, I'm not sure if that's where we want it. We might eventually want it to be attached to an axis e.g. how an `xaxis` or `yaxis` is currently. @campagnola WDYT?

bollu commented:
yep, I realize. I just wanted something quick-and-dirty to start off with.

I think we would position the colorbar based on the orientation given. If it's oriented vertically, then it would be positioned to the left. Horizontally, then to the bottom.

bollu commented:
Also, it needs to be independent of scaling, right?

larsoner commented:
How do you mean?

bollu commented:
Right now, when you zoom in / out, the colorbar get scaled up or down. But that's not supposed to happen, since it's a part of the dataset, right? So it should be free from the scaling, the same way the axis don't scale, but just move around.

larsoner commented:
I could see it going either way. Maybe it should scale up and down with the data if it is vertical, and left/right if it is horizontal, but with a fixed colorbar "width" in either case.

bollu commented:
Yeah, I suppose that works too.

Another option would be rather than scaling the colorbar, we could implement the "cutting in" that we had discussed before. So the scale remains the same, but the `clim` and the color-mapping stretches or cuts in to the range based on the data on-screen?

That may get confusing, however.

bollu committed:
change padding algorithm

bollu committed:
cleanup spaces

bollu commented:
so, I've been messing about with this. The `TextVisual` doesn't seem to obey the scaling induced by the `AxisWidget`. I spent ~an hour looking for bugs in `ColorBar` until I decided to manually attach a `TextVisual` onto a `PlotWidget`.

The `TextVisual` just doesn't scale.

@campagnola - Do you have any idea why this could happen? I'm looking into how the transforms trickle down right now

bollu commented:
minimal gist to reproduce: [link](https://gist.github.com/bollu/7efac08f48fca8e38601)

larsoner commented:
The text stays a fixed size, the way to change it is to change the point
size. Why do you want it to scale? I think in most cases it shouldn't.

larsoner commented:
For example, the axis labels also don't scale with window size, and I think
that's the correct behavior

bollu committed:
replace border rendering from shader to using PolygonVisual

NOTE: this is purely a stopgap arrangement till I figure
out why the borders aren't getting clipped inside a PlotWidget

bollu commented:
Because the `ColorBarVisual` uses the text for labels and limits - that should scale when the actual `colorbar` scales.

bollu commented:
Also, this was maybe stupidity, but I wanted to test it out:

The old `ColorBarVisual` _border_ doesn't get clipped by the Axes -- 

![screenshot from 2015-07-19 01-02-39](https://cloud.githubusercontent.com/assets/1694861/8763271/f4b1a89a-2db1-11e5-9514-f82f353aaa08.png)

So, I changed it to use the PolygonVisual to render borders just to see if it works, and it does:

![screenshot from 2015-07-19 00-58-50](https://cloud.githubusercontent.com/assets/1694861/8763287/596ad8ec-2db2-11e5-8a73-0bd4f3c79ecc.png)

I'll have to debug it to see why it doesn't work.
@Eric89GXL do you know how the Axes discard polygons? do they use `glScissor`? or something else?

larsoner commented:
I don't think the text should scale up, the labels are a lot like axis
labels which also do not scale.

I believe the clipping is done in the fragment shader now. I'm surprised it
doesn't work. It suggests there is some underlying problem with the visual
that draws the border, or in the system that sets the clipping in the
shaders.

bollu commented:
@Eric89GXL the shader doesn't do anything special - it just "inflates" tri strips along the correct direction given a border width. But I'll check.

campagnola commented:
Clipping works by inserting a function call into the end of the fragment shader (using a Clipper filter), so perhaps the old colorbar was implemented in such a way that the filter attachment was not handled correctly.

bollu commented:
@campagnola: hm, could you give me a pointer on how to handle filters?

campagnola commented:
The Visual base class generally takes care of this, and CompoundVisual should take care of forwarding filter attachments to its internal sub-visuals. If it's not working correctly, you can use `--vispy-log=debug` to see the complete shader code. You should see a call to the clipping function inserted at the end of the fragment shader's main() function.

bollu commented:
@Eric89GXL So, in that case, should the `colorbar` be rendered "outside" the Axis grid? (so that it's size is constant?)

bollu commented:
@campagnola ah, the shader that draws the border is a separate shader, and the `CompoundVisual` wouldn't know of it. I'm guessing I'll need to use a `MultiProgram` for this?

bollu commented:
@campagnola EDIT - I meant the `Visual` class which the `CoreColorBar` inherits from would be aware of the second shader, so the filters would not be set for it.

bollu committed:
make borders clip by separating border rendering into another visual.

NOTE: The method use is pretty much "divide and conquer". I would love
it if there was a nicer way to do this than simply convert everything
that needs its own transforms into a separate Visual. However, the
method works and also, interestingly enough, improves code readability
as now every mutation is in its own class

bollu commented:
So, I split up the border rendering into _another_ Visual which belongs only to `colorbar.py`.

#### Pros:
1. Code is neater, since state changes are isolated.
2. Separation of concerns is nice.

#### Cons:
1. too many `Visual`s?
2. I'm pretty sure there's a nicer way to structure this. Thoughts, guys?

#### Screenshot:

![screenshot from 2015-07-21 02-37-10](https://cloud.githubusercontent.com/assets/1694861/8787463/d8e34a78-2f51-11e5-838f-715b681268d9.png)

bollu committed:
fix flake, remove dead code.

larsoner commented on vispy/visuals/colorbar.py:
```diff
@@ -96,6 +96,105 @@
+    @border_color.setter
+    def border_color(self, border_color):
+        self._border_color = Color(border_color)
+        self.shared_program.frag['border_color'] = self._border_color.rgba
```
Can we refactor this with the visual that draws the borders for e.g. the ViewBox classes? Not sure if there is enough shared code to make it worthwhile or not.

bollu replied:
Luke mentioned this too, that the `ViewBox` also renders borders. I'll take a look at it. Maybe in a separate PR?

larsoner replied:
Sure if you want, but this PR already looks like it has all the code you'd need to plug into `ViewBox`. If it means moving this code to some other file like `visuals/border.py` and then importing something both here and in `viewbox.py`, I think it's fine to do in this PR. It just means 90% red changes in `viewbox.py` (or wherever it is -- `widget.py`?) hopefully.

bollu committed:
fix bug with rendering. The anchors weren't set right, and the
coordinate systems considered were wrong.

TODO: add more padding based on the border width.
NOTE: border width is in px. This should be converted to the visual
coordinate system.

bollu committed:
fix labels rendering: take border width into account as well

bollu committed:
Splitting up colorbar into BorderVisual, rewriting ColorBar to work with
transforms correctly

bollu commented:
Quick update - 

I moved the border rendering of the `ColorBarVisual` into a `BorderVisual` class in a separate file, which will be used with the `Widget` for border rendering.

I realized when I was rewriting the `ColorBarVisual` that I was placing text without taking the correct coordinate system into account. So,  now, I'm going from `visual -> document`, moving vectors in that coordinate system (to place the label and the end ticks), then transforming back `document -> visual`.

bollu commented:
Hm, on looking at the `Widget` code, I see that the `MeshVisual` is used by the widget to set the background color as well as draw the borders. So, replacing the `MeshVisual` doesn't seem.. prudent?

I think we should keep both the `MeshVisual` to setup things like background color, and the `BorderVisual` to draw borders of different colors and widths.

Thoughts?

larsoner commented:
Agreed. I forgot that used `MeshVisual` under the hood.

bollu commented on vispy/visuals/visual.py:
```diff
@@ -649,7 +649,7 @@ def _compute_bounds(self, axis, view):
                 if bounds is None:
                     bounds = vb
-                else:
+                elif vb is not None:
```
This was a bug that I caught: the `CompoundVisual` would access the bounds of the sub-visual, but wouldn't check if the sub-visual returned `None` as its bounds.

bollu replied:
@campagnola: Can you could confirm that this is a bug and not intended behavior?

bollu committed:
add accessors for border_width, border_color to be used by ViewBox

bollu committed:
get rudimentary borders working

bollu committed:
borders work; width, color can be modified

bollu commented:
@Eric89GXL - I integrated the `BorderVisual` into `Widget`. I'm off to bed now, so I'll take a closer look tomorrow morning. I believe its working now, though!

larsoner commented:
Looks reasonable to me. We will lose a little bit of performance by adding more visuals since it's more programs to run, but we will eventually need to address that problem systematically, e.g. by collectionizing somehow. So looks reasonable to me for now.

bollu commented:
Did you test it out as well? Or just check the code out?

bollu committed:
implemented the new system for 'bottom' orientation

bollu commented:
@Eric89GXL: How do anchors behave on transformation? Like, on the "current" commit, the example at `examples/basics/visuals/colorbar_visual_types.py` renders the `ColorBar` with the text at the correct position. 

However, plugging the same `ColorBar` into a scenegraph messes the text rendering up. I finally decided to just attach a turntable camera and mess about the `ColorBar` at different angles.

It looks to me like the anchors are invariant under rotations, which means that text that renders "fine" in one orientation is wrong in the 180 degree case.

[here is the gist I was using to test](https://gist.github.com/bollu/f8b1c06929e3c1997d3c)

#### Correct version in `colorbar_visual_types.py` (bottom orientation):

![example_correct_orientation](https://cloud.githubusercontent.com/assets/1694861/8859824/652f575e-319d-11e5-8fa2-4c419e7a240c.png)

#### Wrong orientation in scenegraph:

![scenegraph_wrong_orientation](https://cloud.githubusercontent.com/assets/1694861/8859835/77f31a06-319d-11e5-936b-67059cf5d8fd.png)

#### Correct orientation in same scenegraph example, just rotated 180 degrees:

![scenegraph_correct_orientation](https://cloud.githubusercontent.com/assets/1694861/8859837/7b763334-319d-11e5-87e5-acc002e86640.png)

bollu committed:
rewrite positioning code of labels in ColorBar

bollu committed:
make borders thicker in colorbar_visual_types so it is more obvious that the labels are placed right

bollu commented on vispy/visuals/colorbar.py:
```diff
@@ -436,76 +314,125 @@ def _update(self):
-            self._ticks[0].pos = x - halfw, text_y
-            self._ticks[1].pos = x + halfw, text_y
+        (label_pos, ticks_pos) = \
+            ColorBarVisual._calc_positions(center=self._center_pos,
```
I created a static member which calculates the label positions. I prefer stuff that does pure computation to be static, since that reinforces the fact that it doesn't exactly depend on the class's state.

bollu commented on vispy/visuals/colorbar.py:
```diff
@@ -436,76 +314,125 @@ def _update(self):
+        #               v
+        #               .(center)
+        # <-half_axis_x->
+        # --------------------
```
Added an ASCII-picture describing the coordinate systems involved. Can remove if it's against guidelines

bollu commented on vispy/visuals/colorbar.py:
```diff
@@ -436,76 +314,125 @@ def _update(self):
             return "center", "top"
 
+    @staticmethod
+    def _calc_positions(center, halfdim, border_width,
```
All the ugly branching calculations in handled in one function, so yay, neater code :)

bollu commented on vispy/visuals/colorbar.py:
```diff
@@ -436,76 +314,125 @@ def _update(self):
+        perp_scale = np.linalg.norm(perp_axis)
+        perp_scale += border_width
+        perp_scale *= _TEXT_PADDING_FACTOR
+        perp_axis *= perp_scale / np.linalg.norm(perp_axis)
```
@Eric89GXL - I'm calculating the position as you suggested: figure out the distance, add the border width, then scale by a small amount (`_TEXT_PADDING_FACTOR`)

bollu commented:
Update:

I rewrote the position calculations to use the correct coordinate systems. The code also got nicer to read thanks to the changes involved.

I'm not sure about the anchors (the question I posted about anchors under transformation). Once that's resolved, I should be able to get it rendering properly under every transform.

I'm reluctant to start writing the `PlotWidget` code for `ColorBar` while I'm unsure about how anchors work,

bollu committed:
adapted AxisVisual code for placing labels

larsoner commented:
Same with this, can you close and open against `vispy/scenegraph-update`?

bollu committed:
added comment explaining weird vectors for perp

bollu committed:
thicken borders in example to test out placement

bollu committed:
fix error, change direction -> perp_direction

bollu commented:
Sure thing. Although, won't Travis break because of unrelated linting errors?

larsoner commented:
How do you mean? I'm going to work on `scenegraph-update` so that it doesn't have any `flake` errors, and hopefully you're producing flake-error-free code :)

bollu commented:
Haha, like, I _think_ that the version of `scenegraph-update` this was forked from had a few linting errors from unrelated modules. So, since you've fixed them, I guess I'll rebase?

larsoner commented:
Yeah that should work. But you might not even need a rebase, Travis will test the merged version
